### PR TITLE
[7.0.x] forward port changes from 6.3.x and 5.5.x

### PR DESCRIPTION
**Pull request by a-palchikov**
This PR catches up with 6.3.x and 5.5.x branches and ports the following additionally:
https://github.com/gravitational/gravity/pull/326
https://github.com/gravitational/gravity/pull/1081
https://github.com/gravitational/gravity/pull/1043
https://github.com/gravitational/gravity/pull/1093
https://github.com/gravitational/gravity/pull/1278

Requires https://github.com/gravitational/gravity.e/pull/4294.

### CODE REVIEW

File: lib/ops/opsservice/agents.go [hunk 1]
```diff
@@ -588,16 +599,29 @@ func (r *agentGroup) add(p rpcserver.Peer, hostname string) {
...
 }
 
 // hasPeer determines whether the group already has a peer with the specified
-// hostname but a different address
+// address and hostname
 func (r *agentGroup) hasPeer(addr, hostname string) bool {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	for existingAddr, existingHostname := range r.hostnames {
+		if existingHostname == hostname && existingAddr == addr {
```
Comment (r0mant) [inline lib/ops/opsservice/agents.go:615]: Hmm, looking at how this is used, won't this allow the case where you launch "gravity join" twice on the same host?

File: tool/gravity/cli/system.go [hunk 2]
```diff
@@ -47,6 +51,38 @@ import (
 	"github.com/sirupsen/logrus"
 )
 
+func systemClusterInfo(env *localenv.LocalEnvironment) error {
```
Comment (r0mant) [inline tool/gravity/cli/system.go:54]: What's this used for?

Comment (a-palchikov) [reply tool/gravity/cli/system.go:54]: This wraps `kubectl cluster-info dump --output-directory=/path/to/dir` [here](https://github.com/gravitational/gravity/blob/dmitri/7.0.x/forward-port-6.3.x/lib/report/k8s.go#L41)

File: lib/ops/opsservice/agents.go (hunk 1 above)
Comment (a-palchikov) [reply lib/ops/opsservice/agents.go:615]: Yes, but it should not be possible since the port will be in-use by only one?

**Comment by a-palchikov** [comment]
Superseded by #1361

**Pull request closed without merging**